CASSANDRA-18781 Add guardrail to be able to disable bulk load of SSTables

Runtian opened the pull request:
```
Add config to be able to disable bulk load

```

The [Cassandra Jira](https://issues.apache.org/jira/browse/CASSANDRA-18781)

Runtian committed:
Add ability to disable bulk load

smiklosovic commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -397,6 +397,8 @@ public static class SSTableConfig
     @Replaces(oldName = "max_mutation_size_in_kb", converter = Converters.KIBIBYTES_DATASTORAGE, deprecated = true)
     public DataStorageSpec.IntKibibytesBound max_mutation_size;
 
+    public boolean bulk_load_enabled = true;
```
@Runtian you should introduce and document this property in `cassandra.yml` as well

smiklosovic replied:
also, I think this variable should be `volatile`.

smiklosovic commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -3974,6 +3974,17 @@ public static void setGCWarnThreshold(int threshold)
+    }
+
+    @VisibleForTesting
+    public static void setBulkLoadEnabled(boolean bulk_load_enabled)
```
do not use snake-case here, just use camel-case, `boolean bulkLoadEnabled`

smiklosovic commented on src/java/org/apache/cassandra/streaming/StreamDeserializingTask.java:
```diff
@@ -106,6 +107,11 @@ public StreamSession deriveSession(StreamMessage message)
 
+        if (streamSession.getStreamOperation() == StreamOperation.BULK_LOAD && !DatabaseDescriptor.isBulkLoadEnabled())
+        {
+            throw new RuntimeException("Bulk Load is disabled");
```
`Bulk load is disabled`.

smiklosovic commented on conf/cassandra.yaml:
```diff
@@ -1889,6 +1889,9 @@ drop_compact_storage_enabled: false
 # Guardrail to allow/disallow DROP KEYSPACE statements
 # drop_keyspace_enabled: true
 #
+# Guardrail to allow/disallow BULK LOAD
```
why do you use capitals? Why is not "bulk load" enough?

Guardrail to allow/disallow bulk loading of SSTables

We should mention what we actually want to bulk load ... bulk load of what?

smiklosovic commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -397,6 +397,8 @@ public static class SSTableConfig
     @Replaces(oldName = "max_mutation_size_in_kb", converter = Converters.KIBIBYTES_DATASTORAGE, deprecated = true)
     public DataStorageSpec.IntKibibytesBound max_mutation_size;
 
+    public volatile boolean bulk_load_enabled = true;
```
There is a section down below in this file where all guardrail configuration fields are specified. Could you move it down there so they are all logically coupled together?

smiklosovic commented on src/java/org/apache/cassandra/db/guardrails/Guardrails.java:
```diff
@@ -184,6 +184,15 @@ public final class Guardrails implements GuardrailsMBean
                     "DROP KEYSPACE functionality");
 
+    /**
+     * Guardrail disabling BULK LOAD behavior
```
Guardrail disabling bulk loading of SSTables

smiklosovic commented on src/java/org/apache/cassandra/db/guardrails/Guardrails.java:
```diff
@@ -184,6 +184,15 @@ public final class Guardrails implements GuardrailsMBean
+    new EnableFlag("bulk_load_enabled",
+                   null,
+                   state -> CONFIG_PROVIDER.getOrCreate(state).getBulkLoadEnabled(),
+                   "BULK LOAD functionality");
```
"bulk load", maybe we can be more descriptive here like:

"bulk loading of SSTables"

smiklosovic commented on src/java/org/apache/cassandra/db/guardrails/GuardrailsConfig.java:
```diff
@@ -175,6 +175,13 @@ public interface GuardrailsConfig
     boolean getDropKeyspaceEnabled();
 
+    /**
+     * Returns whether BULK LOAD is allowed
```
bulk load of SSTables

smiklosovic commented on src/java/org/apache/cassandra/db/guardrails/GuardrailsMBean.java:
```diff
@@ -311,6 +311,18 @@ public interface GuardrailsMBean
     boolean getDropKeyspaceEnabled();
 
+    /**
+     * Sets whether BULK LOAD is allowed
```
same and next method again

smiklosovic commented on test/unit/org/apache/cassandra/streaming/async/StreamingInboundHandlerTest.java:
```diff
@@ -105,13 +108,53 @@ public void StreamDeserializingTask_deriveSession_StreamInitMessage()
+            // verify that other operation is not affected
+            StreamInitMessage msg = new StreamInitMessage(REMOTE_ADDR, 0, nextTimeUUID(), StreamOperation.REPAIR, nextTimeUUID(), PreviewKind.ALL);
+            StreamDeserializingTask task = new StreamDeserializingTask(null, streamingChannel, MessagingService.current_version);
+            StreamSession session = task.deriveSession(msg);
```
can be done like `Assert.assertNotNull(task.deriveSession(msg))`

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,144 @@
+/**
+ * Tests the guardrails for bulk load, {@link Guardrails#bulkLoadEnabled}.
+ */
+public class GuardrailBulkLoadEnabledTest extends GuardrailTester
```
You are not testing that enabled bulk loading is loading sstables and that we can enable and disable (go back and forth) You only test that disabled bulk loading is not loading them.

I would do:

1) start a node
2) create tables, add data, backup sstables, truncate
3) bulk load sstables and verify data are there, truncate again
4) disable bulk load (by calling a respective method on `Guardrails` on server side)
5) bulk load sstables and verify that it is not loading it
6) enable bulk load by same mean as in 4)
7) load sstables and verify it is loading it

4-7 is important as we know that we can actually go with enable -> disable -> enable so we can alter this configuration and it works as expected.

smiklosovic commented on src/java/org/apache/cassandra/streaming/StreamDeserializingTask.java:
```diff
@@ -109,6 +111,12 @@ public StreamSession deriveSession(StreamMessage message)
         // in all other cases, no new control channel will be added, as the proper control channel will be already attached.
         streamSession.attachInbound(channel);
+
+        if (streamSession.getStreamOperation() == StreamOperation.BULK_LOAD && !Guardrails.bulkLoadEnabled.isEnabled(null))
```
We tend to omit braces when bodies of `if`s contain just one statement. In this case you do not need braces here.

smiklosovic commented on test/unit/org/apache/cassandra/streaming/async/StreamingInboundHandlerTest.java:
```diff
@@ -105,13 +108,53 @@ public void StreamDeserializingTask_deriveSession_StreamInitMessage()
+    }
+
+    @Test
+    public void StreamDeserializingTask_deriveSession_BulkLoad_Disabled_Do_Not_Affect_Non_Bulk_Streaming() throws Exception
```
should be actually `Disabled_Does_Not_Affect`

Do_Not -> Does_Not

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,144 @@
+    @Test
+    public void bulkLoaderFailedWithBulkLoadDisabled() throws Throwable
+    {
+        File sstables_to_upload = prepareSstablesForUpload();
```
Why are you using snake case? Should be camel case as everywhere, there are multiple instances of this happening in this file.

smiklosovic commented:
@Runtian the current code does not pass `ant check`. Your code contains unused imports both in production as well as in tests.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,171 @@
+
+    private static File copySstablesFromDataDir(String table) throws IOException
+    {
+        File cfDir = new File("build/test/cassandra/bulk_load_tables", table);
```
@Runtian please move this to a temporary directory. Also, if we are going to run this test in a multiplexer (e.g. 500 times in a loop to see if a test is stable), you are not deleting this directory after the test. That means that we are polluting the disk space unnecessarily.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,171 @@
+    private static File prepareSstablesForUpload() throws IOException
+    {
+        generateSstables();
+        File sstable_dir = copySstablesFromDataDir("test");
```
@Runtian this is still on snake case

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,171 @@
+    {
+        CLUSTER = init(Cluster.build(1)
+                              .withConfig(c -> {
+                                  c.with(Feature.NATIVE_PROTOCOL, Feature.NETWORK, Feature.GOSSIP);
```
could you statically import these enums? You also do not need to wrap this in `{` and `}` as that closure is a one liner.

If you do all of that, you realize that whole  `build(1).withConfig(c -> c.with(...).start()` can fit on one line below 120 characters. You might statically import `build` method as well.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,171 @@
+    {
+        File sstablesToUpload = prepareSstablesForUpload();
+        // bulk load SSTables work as expected
+        ToolRunner.ToolResult tool = ToolRunner.invokeClass(BulkLoader.class,
```
I would create a helper method for this. You are calling it 3 times with same parameters.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,171 @@
+    {
+        File cfDir = new File("build/test/cassandra/bulk_load_tables", table);
+        cfDir.tryCreateDirectories();
+        List<String> keyspaceDirPaths = CLUSTER.get(1).callOnInstance(() -> {
```
you dont need `transform` for this:

        CLUSTER.get(1).callOnInstance(() -> Keyspace.open("bulk_load_tables")
                                                               .getColumnFamilyStore("test")
                                                               .getDirectories()
                                                               .getCFDirectories()
                                                               .stream()
                                                               .map(File::absolutePath)
                                                               .collect(toList()));

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,171 @@
+        {
+            for (File file : srcDir.tryList((file) -> file.isFile()))
+            {
+                FileUtils.copyFileToDirectory(file.toJavaIOFile(), cfDir.toJavaIOFile());
```
the rule of "one liner does not need to be wrapped in braces" holds for `for`s too.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,171 @@
+        });
+        for (File srcDir : transform(keyspaceDirPaths, (p) -> new File(p)))
+        {
+            for (File file : srcDir.tryList((file) -> file.isFile()))
```
`tryList(File::isFile)`

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,171 @@
+            CLUSTER.get(1).executeInternal(String.format("INSERT INTO bulk_load_tables.test (pk, val) VALUES (%s, '%s')", i, i));
+        }
+        CLUSTER.get(1).runOnInstance(rethrow(() -> StorageService.instance.forceKeyspaceFlush("bulk_load_tables",
+                                                                                              ColumnFamilyStore.FlushReason.UNIT_TESTS)));
```
If you statically import `UNIT_TESTS` you might fit it on one line.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,171 @@
+        return cfDir;
+    }
+
+    private static void generateSstables() throws IOException
```
This method actually does not throw `IOException`. It does not need to throw anything.

smiklosovic replied:
could you rename this to`generateSSTables` ?

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,171 @@
+        return sstable_dir;
+    }
+
+    private static File copySstablesFromDataDir(String table) throws IOException
```
`copySStablesFromDataDir`

adelapena commented on src/java/org/apache/cassandra/config/GuardrailsOptions.java:
```diff
@@ -364,6 +364,19 @@ public void setDropKeyspaceEnabled(boolean enabled)
                                   x -> config.drop_keyspace_enabled = x);
     }
 
+    public boolean getBulkLoadEnabled()
```
Nit: add `@Override`

adelapena commented on src/java/org/apache/cassandra/db/guardrails/Guardrails.java:
```diff
@@ -184,6 +184,15 @@ public final class Guardrails implements GuardrailsMBean
+     */
+    public static final EnableFlag bulkLoadEnabled =
+    new EnableFlag("bulk_load_enabled",
+                   null,
```
Can we add a reason for disabling bulk loading instead of `null`?

Runtian replied:
Do you have any suggestions on the reason for this feature?

adelapena replied:
That should be the reason why this guardrail exists, to be displayed in the error message. It should be something of the form `Bulk loading is risky because in X conditions Y might happen`, or something like that.

Runtian replied:
This is different from other features as this is disabling the bulk load streaming feature. There is no client state for this and the error message won't be displayed.

adelapena replied:
With the last changes that use `ensureEnabled` the error message with the reason is included in the message, isn't it?

Runtian replied:
I mean the reason won't be sent back to the "client" because there is not client state. Client will only get the "stream failed" message.

Runtian replied:
The reason for disabling bulk load is solely the decision of the Cassandra operator. In our situation, we want to activate this feature only for users who truly require it.

smiklosovic replied:
@Runtian you just wrote the message which should be present in an exception thrown when it is disabled :)

`Bulk loading of SSTables was disabled by Cassandra operator. We activate this feature only for users who truly require it to not make nodes potentially unstable.`

Something like that ... 

Is not it true that guardrail's exception will be propagated to StreamSession#onError? You can probably append `t.getMessage()` via SessionFailedMessage

        if (channel.connected())
        {
            state(State.FAILED); // make sure subsequent error handling sees the session in a final state 
            channel.sendControlMessage(new SessionFailedMessage()).awaitUninterruptibly();
        }

`new SessionFailedMessage(t)`

and then append it in toString when it is not null.

Is @adelapena cool with that?


However, I am not sure how that would work with serialization / deserialization logic. It would probably not work ...

But I think that description is still valuable to have. Maybe in future all these guardrails will be present somewhere for a user so he might see what each guardrail does and what it is good for.

adelapena commented on src/java/org/apache/cassandra/db/guardrails/GuardrailsMBean.java:
```diff
@@ -311,6 +311,18 @@ public interface GuardrailsMBean
+    /**
+     * Sets whether bulk load of SSTables is allowed
+     */
+    void setBulkLoadEnabled(boolean enabled);
```
The setter/getters for `bulk_load_enabled` are inserted between the getters and setters for `drop_keyspace_enabled`, setters and getters for the same property should remain together. Also, on this file we usually write first the getter, then the setter.

adelapena commented on src/java/org/apache/cassandra/db/guardrails/Guardrails.java:
```diff
@@ -184,6 +184,15 @@ public final class Guardrails implements GuardrailsMBean
+    new EnableFlag("bulk_load_enabled",
+                   null,
+                   state -> CONFIG_PROVIDER.getOrCreate(state).getBulkLoadEnabled(),
+                   "bulk loading of SSTables");
```
```suggestion
                   "Bulk loading of SSTables");
```

adelapena commented on src/java/org/apache/cassandra/streaming/StreamDeserializingTask.java:
```diff
@@ -109,6 +110,10 @@ public StreamSession deriveSession(StreamMessage message)
         // in all other cases, no new control channel will be added, as the proper control channel will be already attached.
         streamSession.attachInbound(channel);
+
+        if (streamSession.getStreamOperation() == StreamOperation.BULK_LOAD && !Guardrails.bulkLoadEnabled.isEnabled(null))
+            throw new StreamReceiveException(streamSession, new RuntimeException("Bulk load is disabled"));
```
This isn't a proper usage of guardrails. They should be checked with `ensureEnabled()`, so they throw the formatted `GuardrailViolatedException` defined in their constructor. That would also trigger other associated behaviours, like emitting a log message and a diagnostic event.

Instead, the usage of `EnableFlag#isEnabled` is ignoring all that and using it as a regular boolean config property.

An issue here might be that we don't have a `ClientState`, and guardrails don't throw an exception but a warning on `null` client states: https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/db/guardrails/Guardrail.java#L139-L140

I think we could add a new `Guardrail#throwOnNullClientState()` method to instruct guardrails to throw an exception when triggered with a null `ClientState`. We usually don't throw on null client states to not break background processes such as compaction, but it can make sense for unauthenticated tools like this one.

Runtian replied:
Made the change please review it again. Thanks.

smiklosovic replied:
@adelapena thank you for catching this! It is easy to overlook the difference between `isEnabled` and `ensureEnabled` without deeper look when doing a review just from a browser. On face value it looks like it does the correct job, even returns `boolean`.

Runtian commented:
@smiklosovic I moved the check part out of the `deriveSession` function so we always have a session when throwing error. I removed the unit test since we have the jvm dtest to test this feature and it is not easy to test if it is moved out of the `deriveSession` function.

adelapena commented on src/java/org/apache/cassandra/db/guardrails/EnableFlag.java:
```diff
@@ -117,13 +117,33 @@ public void ensureEnabled(@Nullable ClientState state)
+     * @param throwOnNullClientState If this is true, an exception will be thrown even if client state is null. This is
+     *                               mainly used for features not assoiate with client state such as bulk load of SSTables.
+     */
+    public void ensureEnabled(String featureName, @Nullable ClientState state, boolean throwOnNullClientState)
```
I think this boolean flag shouldn't be added in an overload of `ensureEnabled`, but set once with a method analogous to `Guardrail#minNotifyIntervalInMs(long)`, probably `Guardrail#throwOnNullClientState(boolen)`. 

Same as `minNotifyIntervalInMs`, that method would be called once in the `Guardrails` class, so it's defined once for the guardrail instance.

Adding that on the `Guardrail` class would have the benefit of working for all its subclasses, without the need to modify them with additional method overloads.

Runtian replied:
sure, makes sense.

smiklosovic commented on src/java/org/apache/cassandra/db/guardrails/Guardrails.java:
```diff
@@ -189,10 +189,15 @@ public final class Guardrails implements GuardrailsMBean
                    state -> CONFIG_PROVIDER.getOrCreate(state).getBulkLoadEnabled(),
                    "Bulk loading of SSTables");
 
+    static
```
@Runtian I dont like this. Can't you just call that method as you instantiate `bulkLoadEnabled`? Like this:

    new EnableFlag("bulk_load_enabled",
                   null,
                   "Bulk loading of SSTables might potentially destabilize the node.",
                   state -> CONFIG_PROVIDER.getOrCreate(state).getBulkLoadEnabled(),
                   "Bulk loading of SSTables").throwOnNullClientState(true);

That method returns `this`. There is no reason to not call it on that.

adelapena replied:
+1, although it would require a trivial cast to `EnableFlag` since `Guardrail#throwOnNullClientState` returns a `Guardrail`:
```java
public static final EnableFlag bulkLoadEnabled =
(EnableFlag) new EnableFlag("bulk_load_enabled",
                            "Bulk loading of SSTables might potentially destabilize the node.",
                            state -> CONFIG_PROVIDER.getOrCreate(state).getBulkLoadEnabled(),
                            "Bulk loading of SSTables").throwOnNullClientState(true);
```

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,179 @@
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Tests the guardrails for bulk load, {@link Guardrails#bulkLoadEnabled}.
```
```suggestion
 * Tests the guardrail for bulk load, {@link Guardrails#bulkLoadEnabled}.
```

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,179 @@
+{
+    private static Path tempDir;
+    private static Cluster CLUSTER;
+    static String NODES;
+    static int NATIVE_PORT;
+    static int STORAGE_PORT;
```
Nit: These aren't constants, so they should be camelCased instead of upper case. They can also be `private`.

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/guardrails/GuardrailBulkLoadEnabledTest.java:
```diff
@@ -0,0 +1,179 @@
+                      .map(File::absolutePath)
+                      .collect(toList())
+        );
+        for (File srcDir : transform(keyspaceDirPaths, (p) -> new File(p)))
```
```suggestion
        for (File srcDir : transform(keyspaceDirPaths, File::new))
```

adelapena commented on conf/cassandra.yaml:
```diff
@@ -1889,6 +1889,9 @@ drop_compact_storage_enabled: false
 # drop_keyspace_enabled: true
 #
+# Guardrail to allow/disallow bulk load of SSTables
+# bulk_load_enabled: true
```
Can we add a note about the new guardrail in `NEWS.txt`? It seems it would be the very first new feature in 5.1 :)

It could look like:
```
5.1
===

New features
------------
    - New Guardrails added:
      - Whether bulk loading of sstables is allowed.
```

smiklosovic replied:
OK so I guess the decision to put this to 5.1 was done? It would be great to have it in 5.0 but I can live with this.

This can be done on commit. @Runtian no worries about that.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/SSTableLoaderEncryptionOptionsTest.java:
```diff
@@ -49,16 +49,16 @@
-    static int NATIVE_PORT;
-    static int STORAGE_PORT;
-    static int SSL_STORAGE_PORT;
+    static Cluster cluster;
```
@Runtian I think you fixed the wrong test. `GuardrailBulkLoadEnabledTest` is untouched when it comes to these changes. Would you mind to keep this one as it was?

Runtian replied:
oops, let me fix this.
